src/sparse_blas/backends/cusparse/operations/cusparse_spsv.cpp

```diff
@@ -0,0 +1,249 @@
+}
+
+inline void common_spsv_optimize(oneapi::mkl::transpose opA, bool is_alpha_host_accessible,
+                                 matrix_view A_view, matrix_handle_t A_handle,
+                                 dense_vector_handle_t x_handle, dense_vector_handle_t y_handle,
+                                 spsv_alg alg, spsv_descr_t spsv_descr) {
+    check_valid_spsv("spsv_optimize", A_view, A_handle, x_handle, y_handle,
+                     is_alpha_host_accessible);
+    if (!spsv_descr->buffer_size_called) {
+        throw mkl::uninitialized("sparse_blas", "spsv_optimize",
+                                 "spsv_buffer_size must be called before spsv_optimize.");
+    }
+    spsv_descr->optimized_called = true;
+    spsv_descr->last_optimized_opA = opA;
+    spsv_descr->last_optimized_A_view = A_view;
+    spsv_descr->last_optimized_A_handle = A_handle;
+    spsv_descr->last_optimized_x_handle = x_handle;
+    spsv_descr->last_optimized_y_handle = y_handle;
+    spsv_descr->last_optimized_alg = alg;
+}
+
+void spsv_optimize_impl(cusparseHandle_t cu_handle, oneapi::mkl::transpose opA, const void *alpha,
+                        matrix_view A_view, matrix_handle_t A_handle,
+                        dense_vector_handle_t x_handle, dense_vector_handle_t y_handle,
+                        spsv_alg alg, spsv_descr_t spsv_descr, void *workspace_ptr,
+                        bool is_alpha_host_accessible) {
+    auto cu_a = A_handle->backend_handle;
+    auto cu_x = x_handle->backend_handle;
+    auto cu_y = y_handle->backend_handle;
+    auto type = A_handle->value_container.data_type;
+    set_matrix_attributes("optimize_spsv", cu_a, A_view);
+    auto cu_op = get_cuda_operation(type, opA);
+    auto cu_type = get_cuda_value_type(type);
+    auto cu_alg = get_cuda_spsv_alg(alg);
+    auto cu_descr = spsv_descr->cu_descr;
+    set_pointer_mode(cu_handle, is_alpha_host_accessible);
+    auto status = cusparseSpSV_analysis(cu_handle, cu_op, alpha, cu_a, cu_x, cu_y, cu_type, cu_alg,
+                                        cu_descr, workspace_ptr);
+    check_status(status, "optimize_spsv");
+}
+
+void spsv_optimize(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                   matrix_view A_view, matrix_handle_t A_handle, dense_vector_handle_t x_handle,
+                   dense_vector_handle_t y_handle, spsv_alg alg, spsv_descr_t spsv_descr,
+                   sycl::buffer<std::uint8_t, 1> workspace) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    if (!A_handle->all_use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    common_spsv_optimize(opA, is_alpha_host_accessible, A_view, A_handle, x_handle, y_handle, alg,
+                         spsv_descr);
+    // Ignore spsv_alg::no_optimize_alg as this step is mandatory for cuSPARSE
+    // Copy the buffer to extend its lifetime until the descriptor is free'd.
```

Comment:
The way `no_optimize_alg` is implemented, it should throw the `oneapi::mkl::unimplemented` exception for cuSPARSE backend. Intel oneMKL supports `no_optimize_alg`, but not cuSPARSE.

Reply:
Sure, I've changed this in https://github.com/oneapi-src/oneMKL/pull/527/commits/9bc77d1548a24b6cf3219c04b8a32e95aeb97218. There were some small complications with `release_spsv_descr` where the descriptor may not contain initialized data handles which was not tested before. This should be good now.

Reply:
I remember just now that we have said in the [spec](https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneMKL/source/domains/spblas/operations/spsv.rst?plain=1#L275):
```quote
The algorithm defaults to ``spsv_alg::default_alg`` if a backend does not support the provided algorithm.
```
The idea was to make sure the API is still easy to use for different backends. Portability is important for oneMKL Interface. We could just improve the documentation to clarify that `no_optimize_alg` still calls `cusparseSpSV_analysis`.

Reply:
I have mostly reverted the change in https://github.com/oneapi-src/oneMKL/pull/527/commits/342380ea5a64a27705faf6367bac25ea52114208

Reply:
In that case, I believe https://github.com/oneapi-src/oneMKL/blob/342380ea5a64a27705faf6367bac25ea52114208/docs/domains/sparse_linear_algebra.rst?plain=1#L144-L146 should be changed to
```
 * - ``csr_alg3`` 
   - none 
   - ``CUSPARSE_SPMV_ALG_DEFAULT``
```

Reply:
> I remember just now that we have said in the [spec](https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneMKL/source/domains/spblas/operations/spsv.rst?plain=1#L275):
> 
> ```
> The algorithm defaults to ``spsv_alg::default_alg`` if a backend does not support the provided algorithm.
> ```
> 
> The idea was to make sure the API is still easy to use for different backends. Portability is important for oneMKL Interface. We could just improve the documentation to clarify that `no_optimize_alg` still calls `cusparseSpSV_analysis`.

I do feel a bit torn about this though. On one hand, we'd like to keep it easy for users to use these algorithm enums across hardware vendors. At the same time, however, we would like to avoid unfair comparisons (e.g, `csr_alg3` that cusparse does not support versus rocsparse backend that does support it) or more importantly, drastic performance changes (read: drops) if/when a backend adds new algorithm support (e.g., cusparse adding support for `csr_alg3`).

The mapping of Spec enums to backend enums is anyway arbitrary, and it is difficult for the spec to define a mapping across backends that respects the algorithm properties/behaviors (e.g., bitwise-reproducibility) across backends. That means that in all likelihood, users caring about cross-vendor portability would need to write code to choose specific, possibly different algorithm enums for each backend.

Overall I'm slightly inclined toward disallowing dispatching to default algorithms when a backend doesn't support a particular `enum`, but I don't have a strong reason for or against it. @spencerpatty, any thoughts on this?

Reply:
The way I see it oneMKL Interface is supposed to have a slightly "higher level" API than its backends and smoothen some of the differences between them so the risks that you mention are bound to happen.
Thanks for the approval. FYI I am planning to get a second approval by the end of the week and merge it. I would also need an approval on the spec PR please! https://github.com/uxlfoundation/oneAPI-spec/pull/583

Reply:
Hmm,  this is a tricky thing and I think the best thing is to have documented fallback (maybe it is default alg) when the particular alg enum requested does not exist, otherwise map  alg1(oneMath Interfaces) to alg1 (backend) etc

The place where this is going to be most complicated is the TRSV case with analysis step sometimes being required and possibly the SpGEMM of the future where sortedness may be a requirement ...